src/item.cpp

```diff
@@ -4179,66 +4179,35 @@ item *item::get_usable_item( const std::string &use_name )
     return nullptr;
 }
 
-item_location item::pick_reload_ammo( player &u, bool interactive ) const
-{
-    std::set<ammotype> ammo_types; // any ammo for empty detachable or integral magazines
-    std::set<itype_id> item_types; // specific ammo for partially loaded detachable or integral magazines
-    std::set<itype_id> compat_mag; // compatible magazine types for items that reload using them
-
-    // checks if item reloadable and if so appends to either ammo_type or item_type
-    auto wants_ammo = [&ammo_types,&item_types,&compat_mag]( const item& it ) {
-        if( !it.has_flag( "NO_RELOAD" ) ) {
-            if( it.magazine_integral() || it.is_magazine() ) {
-                if( it.ammo_remaining() < it.ammo_capacity() ) {
-                    if( it.ammo_remaining() > 0 ) {
-                        item_types.insert( it.ammo_current() );
-                    } else {
-                        ammo_types.insert( it.ammo_type() );
-                    }
-                }
-            } else if( !it.magazine_current() ) {
-                const auto mags = it.magazine_compatible();
-                compat_mag.insert( mags.begin(), mags.end() );
-            }
-        }
-    };
-
-    // consider both the item and any attached gunmods
-    wants_ammo( *this );
-    if( is_gun() ) {
-        std::for_each( contents.begin(), contents.end(), wants_ammo );
+bool item::can_reload() const {
+    if( !is_gun() && !is_tool() && !is_magazine() ) {
+        return false;
+    }
+    if( has_flag( "NO_RELOAD") || has_flag( "RELOAD_AND_SHOOT" ) ) {
+        return false;
+    }
+    if( ammo_type() == "NULL" ) {
+        return false;
+    }
+    if( magazine_integral() && ammo_remaining() >= ammo_capacity() ) {
+        return false;
     }
+    // either intergral magazine with sufficient capacity or uses detachable magazines
```

Comment:
Guns with detachable magazines are always considered reloadable?  If they have a magazine loaded you can't reload without loading first, same as with integral magazines.